### ISSUE TITLE
[Suggestion] Make relation compatible with assumptions module

### ISSUE THREAD

**Issue by mcpl-sympy** [sympy/sympy#20003]
Equality, inequality, and every other binary relations are [finitary relation](https://en.wikipedia.org/wiki/Finitary_relation). Unary predicate, which is finitary relation as well, is implemented as `Predicate` and `AppliedPredicate` in `assumptions` module.  

I suggest that every other finitary relation can follow this design. Equality symbol, for instance, can be implemented as instance of `Relation` (just as map is being implemented as instance in #19953),  and equation can be instance of `AppliedRelation`.

Also, API between assumption module and `AppliedRelation` can be established so that both can enjoy each other's rich functionalities. Here is what I am thinking about:

```python
>>> eq = Equality() # may be implemented as singleton
>>> eqn = eq(x+1, y+1)

>>> srepr(eqn)
AppliedRelation(Equality(), (Add(Symbol('x'), Integer(1)), Add(Symbol('y'), Integer(1)))
>>> print(eqn)
x + 1 = y + 1

>>> print(ask(eqn)) # None, since truth value cannot be determined
None
>>> with assuming(eq(x, y)):
...     print(ask(eqn)) # with assuming x = y, it is true that x+1 = y+1
True
```

Another example:

```python
>>> gt = GreaterThan()
>>> gt(y, 1)
y > 1
>>> with assuming(gt(y, 1)):
...     print(refine(sqrt((y-1)**2)))
y - 1
```

My group is planning to start working on this in next month. Before that, I would like to have some opinions and suggestions.

**Comment by oscarbenjamin** [comment]
I don't understand what how this would help with anything.

> Also, API between assumption module and `AppliedRelation` can be established so that both can enjoy each other's rich functionalities.

What does this mean?

**Comment by mcpl-sympy** [comment]
@oscarbenjamin 
Assuming range for a variable is difficult in current SymPy. Suppose we want to write "if 0<x and x<𝜋, sin(x) is positive" with new assumption system. "0<x" is easy - we can use `Q.positive(x)`. But how about "x<𝜋"?  
Perhaps `Q.negative(x - pi)` or `Q.is_true(x < pi)` can represent it, but these seem to be only short-term measurements for the issue. The best way will be to introduce n-ary predicate. With something like `Q.lessthan(x, pi)`, we can elegantly do things like this:
```python
>>> refine(Abs(sin(x)), Q.positive(x) & Q.lessthan(x, pi))
sin(x)
```

Here is another example. How would we express "a and b are coprime" in SymPy?
My answer: we can introduce binary predicate `Q.coprime` and express it as `Q.coprime(a, b)`. With this, we can do number theories like
```python
>>> with assuming(Q.even(a) & Q.even(b)):
...     ask(Q.coprime(a, b))  # return False since even numbers are never coprimes
False
```

In short, I'd say that the goal is to expand current `assumption` module (which is confined to unary predicate) to finitary predicate. Binary relation, which I am planning to focus, will be implemented along the way since they are binary predicate.

**Comment by oscarbenjamin** [comment]
Okay, that makes sense. I guess the hard part is writing an inference system that can actually make sense of the binary predicates.

Being able to handle inequality predicates is a long wanted feature so if we can actually make sense of that in refine/ask then that's great.

**Comment by asmeurer** [comment]
What is the need for having `AppliedRelation(Equals, x, y)` instead of just `Equals(x, y)`? Even if Equals works like predicates on Q so that it can be represented individually, this works just fine with the current design there

```py
>>> Q.positive(x).args
(x,)
>>> Q.positive(x).func
Q.positive
>>> isinstance(Q.positive, Basic)
True
>>> isinstance(Q.positive(x), Basic)
True
```

Anyway, I don't really see this as being the fundamental problem of why relations don't work in the assumptions. So long as `x < b` returns some kind of Boolean object it should be allowable in predicates (`refine(Abs(sin(x)), Q.positive(x) & (x < pi))`). The issue is that it presently doesn't recognize it. It needs to put it in a canonical form one way or another, because `Q.positive(x - y)` and `x > y` are both the same thing, so there's no reason to write the same logic for them both twice. Whether it is simpler to try to put everything into the `Q.positive` form or `x > y` isn't clear to me. I would guess the former is simpler to deal with because it's unique (e.g., `Q.positive(x + y + z)` could be `x + y + z > 0` or `x + y > -z` or `x > -y - z` or ...). So that's just a question of canonicalizing the input predicate expression before passing it off to the handlers. The real problem is that presently the handlers can't handle anything nontrivial. They need to learn about basic predicates about inequalities like transitivity and things like `x > 0 & y > 0 => x*y > 0`, and assuming we use the Q.positive form, how to effectively pattern match them. There's also the sneaking semantic issue which is that inequalities raise exceptions on complex numbers, when it would be simpler to deal with them from a logical point of view in the assumptions if complex inequalities just gave False (then `x > 0` would truly be equivalent to `Q.positive(x)`).

I've discussed the above ideas in the past on other issues, but GitHub issues search isn't great so I'm not finding them. I can't even tell if we actually have a master issue for tracking assumptions on inequalities or not.

**Comment by mcpl-sympy** [comment]
@asmeurer 
Well, although `Q.positive(x).func` indeed returns `Q.positive`, it's real class is `AppliedPredicate`...
```python
>>> Q.positive.__class__
sympy.assumptions.assume.Predicate
>>> Q.positive(x).__class__
sympy.assumptions.assume.AppliedPredicate
```
We can follow this design and override `func` and `args` of applied relation, but we'll need `AppliedRelation` nevertheless.

**Comment by mcpl-sympy** [comment]
Should you happen to find where your old discussion is, please give a link here. It will be greatly valuable.

I wonder how the other CAS programs, such as Mathematica, deal with the assumption system.

**Comment by asmeurer** [comment]
I agree the class design should be the same, if not just reusing the same AppliedPredicate classes. Your example looked like that would be the expression tree. 

> I wonder how the other CAS programs, such as Mathematica, deal with the assumption system.

I don't know about Mathematica, but I think the better way to do assumptions is to use a proper SMT solver.

**Comment by asmeurer** [comment]
> Should you happen to find where your old discussion is, please give a link here. It will be greatly valuable.

https://github.com/sympy/sympy/pull/17392 is a related PR. I can't find the places where I outlined the ideas, but I think I basically reiterated them above. 

For the complexes thing, I did find https://github.com/sympy/sympy/issues/13989 and https://github.com/sympy/sympy/issues/15655. Since https://github.com/sympy/sympy/issues/13989 seems to be basically the issue for it, I will just post there my most recent thoughts.

**Comment by gutow** [comment]
In answer to the following thoughts from @JSS95 received by e-mail about combining with #19479:
```
Recently, I have seen your [PR to implement a new Equation class ](https://github.com/sympy/sympy/pull/19479)equipped with a
symbolic equation derivation system. Coincidentally, we were working on the same class with the same name, structure, and
purpose in our own repository - and I was impressed by your work.

My work, however, has experienced a slight design change in the past few months. I focused on the point that every relation is a
binary predicate, and we have a Predicate class in assumptions module. Therefore I plan to make a new relation as predicates
such as "Q.eq", "Q.gt", etc and make them directly compatible with SymPy's assumption system.  And I was going to implement
the operation between these classes, so that we can allow operation between not only equalities but inequalities as well. (For
example, (x=1) + (y>2) -> (x+y>3) )

Since we are working on the same purpose, I think we can discuss and share our ideas, or hopefully collaborate in a series of
PRs. Please consider my proposal - I will be looking forward to your reply!
```
I think ultimately the combination might be a good idea. However, it is clear to me from discussions here and in #19479 that dealing with relations in general is much more difficult than handing expressions with two sides that are defined as equal. 

1. The first problem is the difference between a definition/assumption of equality and the logical test of equality. As you are discussing coupling this to the assumptions module, I believe what you are working on may address some of this.
2. The equations module as I am working on it intentionally starts with the assumption that the two sides of the equation are equal and can be used to show that it is not true by doing valid algebra on the equation ending in obviously false statements such as `2 = 1`.
3. Getting things to work correctly for general relations such as `>` is much more challenging than getting this to work for `=`.
4. Although not an expert on the status of assumptions in SymPy, my limited experience suggests that the two different assumptions methods need to be better harmonized and combined before a general relations tool can be constructed. I believe @asmeurer and @oscarbenjamin may have more thoughts on this. See also [these 300+ open issues](https://github.com/sympy/sympy/issues?q=is%3Aissue+is%3Aopen+assumptions).

So based on this I suggest that we keep things separate and approach the problem in the following way:
1. Define an equation class for the case of `=`. Which is what I am working on presently. If you have any enhancements that are not part of the general relationship class submit those as pull requests to the [repository in which I am using to work on the equation class](https://github.com/gutow/sympy). Once the equation class is folded into SymPy, then pull requests can be made against SymPy.
2. Within the more general relation class, which you appear to be working on, the equation class should be extended with the relation attributes.
3. Once these are both working and folded into SymPy, we should consider making the equation class a subclass of the relation class. I fear there are some serious project/code architecture issues that will need to be addressed. I suspect I am not skilled enough nor familiar enough with all of SymPy to properly address these architecture issues.

To keep this in perspective, I am a physical chemist interested in this as a tool to facilitate algebraic manipulations for teaching and scientific use. I am more interested in getting something working into the code base quickly than getting the most general implementation possible.

**Comment by JSS95** [comment]
@gutow Thanks for the comment. As far as I know, the problem of previous `Equality` was that it's too prone to evaluated to `True` or `False`. Making relations subclass of `Predicate` can resolve this since it never gets evaluated unless `ask()` is used. Also, we can seamlessly meld it into `assumptions` system without wrapping it with `Q.true`.

By fusing our classes together after we are both finished, I expect that we can get a system that is equipped with powerful symbolic manipulation, and can be perfectly evaluated to boolean value with assumptions when required.

I am currently working on my design in my local. #20656 is the prerequisite of this, so I'll make a PR after that is merged.

**Comment by oscarbenjamin** [comment]
Making an Equation class built on Predicate is an interesting idea that I hadn't considered.

The problem with Equality is that it evaluates to True/False but also that it is a Boolean so it should only really support operations that make sense for Boolean like `&` rather than `+`. Predicate also subclasses Boolean though which I think confuses this.

Equation should have a different "kind" from Boolean I think.
